trunk/user/pdnsd/pdnsd/doc/pdnsd.conf.5.in
.\" Generated automatically from the html documentation by html2confman.pl
.\" 
.\" Manpage for pdnsd.conf (pdnsd configuration file)
.\" 
.\" Copyright (C) 2000, 2001 Thomas Moestl
.\" Copyright (C) 2003, 2004, 2005, 2006, 2007 Paul A. Rombouts
.\" 
.\" This manual is a part of the pdnsd package, and may be distributed in
.\" original or modified  form  under  terms  of  the  GNU  General  Public
.\" License,  as  published by the Free Software Foundation; either version
.\" 3, or (at your option) any later version.
.\" You can find a copy of the GNU GPL in the file COPYING in the source
.\" or documentation directory.
.\" 
.TH PDNSD.CONF 5 "Apr 2012" "pdnsd @fullversion@"
.SH NAME
pdnsd.conf \- The configuration file for pdnsd
.hw config
.SH DESCRIPTION
.PP
This manual page describes the layout of the
.BR pdnsd (8)
configuration file and the available configuration options.
The default location of the file is @sysconfdir@/pdnsd.conf. This may be changed
with the \fB-c\fP command line option.
An example pdnsd.conf comes with the pdnsd distribution in the documentation directory
or in @sysconfdir@/pdnsd.conf.sample.
.SH "FILE FORMAT"
.PP
The configuration file is divided into sections. Each section is prefixed with
the section name and opening curlies ({) and closed with closing curlies (}).
In each section, configuration options can be given in the form

\fIoption_name\fP=\fIoption_value\fP;

Option value may be a string literal, a number, a time specification or a constant.
In previous  versions of pdnsd strings had to be enclosed
in quotes ("), but since version 1.1.10 this is no longer necessary, unless
a string contains a special character such as whitespace, a token that normally starts
a comment, or one of ",;{}\".
Since version 1.2.9 a backslash (\) inside a string is interpreted as an escape character,
so it is possible to include special characters in strings (both quoted or unquoted)
by preceding them with a backslash. Some escape sequences are in interpreted as in the C
programming language, e.g. \t becomes a tab,
\n becomes a new-line control char.
.br
A time specification consists a sequence of digits followed by a one-letter suffix.
The following suffixes are recognized:
s (seconds), m (minutes), h (hours),
d (days) and w (weeks).
If the suffix is missing, seconds are assumed.
If several time specifications are concatenated, their values are added together;
e.g. 2h30m is interpreted as 2*60*60 + 30*60 = 9000 seconds.
.br
Some options take more than one value; in this case, the values are separated with commas.
.br
If you may supply one of a set of possible values to an option, this is noted
in the documentation as
(option1|option2|option3|...)
.br
The constants true|false and yes|no
are accepted as synonyms for the constants on|off.
.br
Comments may be enclosed in /* and */, nested comments are possible. If the
# sign or two slashes (//) appear in the configuration file, everything from
these signs to the end of the current line is regarded as a comment and ignored.
.br
There are examples for nearly all options in the sample config file.

.SS global Section
The global section specifies parameters that affect the overall behaviour of the
server. If you specify multiple global sections, the settings of those later in
the file will overwrite the earlier given values.
.br
These are the possible options:

.TP
.B perm_cache=(\fInumber\fP|off);
Switch the disk cache off or supply a maximum cache size in kB. If the disk
cache is switched off, 8 bytes will still be written to disk.
The memory cache is always 10kB larger than the file cache.
This value is 2048 (2 MB) by default.
.TP
.B cache_dir=\fIstring\fP;
Set the directory you want to keep the cache in.
The default is "@cachedir@"
(unless pdnsd was compiled with a different default).
.TP
.B server_port=\fInumber\fP;
Set the server port. This is especially useful when you want to start the
server and are not root. Note that you may also not specify uptest=ping in
the server section as non-root.
.br
The default port is 53, the RFC-standard one. Note that you should only use
non-standard ports when you only need clients on your machine to communicate
with the server; others will probably fail if the try to contact the server
on the basis of an NS record, since the A record that supplies the address for
(among others) name servers does not have a port number specification.
.TP
.B server_ip=\fIstring\fP;
or
.PD 0
.TP
.PD
.B interface=\fIstring\fP;
Set the IP address pdnsd listens on for requests. This can be useful
when the host has several interfaces and you want pdnsd not to listen on
all interfaces. For example, it is possible to bind pdnsd to listen on
127.0.0.2 to allow pdnsd to be a forwarder for BIND.
The default setting for this option is server_ip=any, which means that
pdnsd will listen on all of your local interfaces.
Presently you can only specify one address here; if you want pdnsd to listen on multiple
interfaces but not all you will have to specify server_ip=any
and use firewall rules to restrict access.
.br
The IP address used to need quotation marks around it, but since version 1.1.10
this is no longer necessary.
.br
If pdnsd has been compiled with both IPv4 and IPv6 support, and you want to
specify an IPv6 address here, then unless pdnsd was compiled to start up in IPv6 mode
by default, you will need to use the \-6 command-line option or
set run_ipv4=off first (see below) in order to ensure that the
IPv6 address is parsed correctly.
.br
If pdnsd is running in IPv6 mode and you specify an IPv4 address here,
it will automatically be mapped to an IPv6 address.
.br
\fINew in version 1.2:\fP You may also give the name of an interface
such as "lo" or "eth0" here, instead of an IP address
(this has been tested on Linux, and may or may not work on other platforms).
pdnsd will not bind to the interface name, but will look up the address of the
interface at start-up and listen on that address. If the address of the interface
changes while pdnsd is running, pdnsd will not notice that. You will need to
restart pdnsd in that case.
.TP
.B outgoing_ip=\fIstring\fP;
or
.PD 0
.TP
.PD
.B outside_interface=\fIstring\fP;
\fINew in version 1.2.9:\fP
Set the IP address of the interface used by pdnsd for outgoing queries.
This can be useful when the host has several interfaces and you want pdnsd
to send outgoing queries via only one of them.
For example, if  pdnsd is running on a host with one interface with IP address
192.168.1.1 connected to the local network, and another with IP address 123.xxx.yyy.zzz
connected to the internet, you may specify server_ip=192.168.1.1
and outgoing_ip=123.xxx.yyy.zzz to enforce that pdnsd only responds
to queries received from the local network, and only sends outgoing queries via
the interface connected to the internet.
.br
The default setting for this option is any, which means that
the kernel is free to decide which interface to use.
Like with the server_ip option, you may also give the name of an
interface here, instead of an IP address.
.TP
.B linkdown_kluge=(on|off);
This option enables a kluge that some people might need: when all servers are
marked down, with this option set the cache is not even used when a query is
received, and a DNS error is returned in any case. The only exception from this
is that local records (as specified in rr and source
sections are still served normally.
In general, you probably want to get cached entries even when the network is down,
so this defaults to off.
.TP
.B max_ttl=\fItimespec\fP;
This option sets the maximum time a record is held in cache. All dns
resource records have a time to live field that says for what period of time the
record may be cached before it needs to be requeried. If this is more than the
value given with max_ttl, this time to live value is set to max_ttl.
This is done to prevent records from being cached an inappropriate long period of time, because
that is almost never a good thing to do. Default is 604800s (one week).
.TP
.B min_ttl=\fItimespec\fP;
This option sets the minimum time a record is held in cache. All dns
resource records have a time to live field that says for what period of time the
record may be cached before it needs to be requeried. If this is less than the
value given with min_ttl, this time to live value is set to min_ttl.
Default is 120 seconds.
.TP
.B neg_ttl=\fItimespec\fP;
This option sets the time that negatively cached records will remain valid in the
cache if no time to live can be determined. This is always the case when whole
domains are being cached negatively, and additionally when record types are cached
negatively for a domain for which no SOA record is known to pdnsd. If a SOA is present,
the ttl of the SOA is taken.
.TP
.B neg_rrs_pol=(on|off|auth|default);
This sets the RR set policy for negative caching; this tells pdnsd under which circumstances
it should cache a record type negatively for a certain domain. off will
turn the negative caching of record types off, on will always add a negative
cache entry when a name server did not return a record type we asked it for, and auth
will only add such entries if the answer came from an authoritative name server for that
domain.
.br
\fINew in version 1.2.8:\fP The default setting will add a negatively cached record
if either the answer was authoritive or the answer indicated the name server had "recursion available"
while the query explicitly requested such recursion.
.br
The preset is "default" (used to be auth).
.TP
.B neg_domain_pol=(on|off|auth);
This is analogue to neg_rrs_pol for whole domain negative caching. It should be safe
to set this on, because I have not seen a caching server that will falsely claim that a
domain does not exist.
.br
The default is auth.
.TP
.B run_as=\fIstring\fP;
This option allows you to let pdnsd change its user and group id after operations that needed
privileges have been done. This helps minimize security risks and is therefore recommended. The
supplied string gives a user name whose user id and primary group id are taken. 
.br
A little more details: after reading the config file, becoming a daemon (if specified) and starting
the server status thread, the main thread changes its gid and uid, as do all newly created threads
thereafter. By taking another uid and gid, those threads run with the privileges of the
specified user.
Under Linux and FreeBSD, the server status thread runs with the original privileges only when the strict_setuid option
is set to off (see below, on by default), because these may be needed
for exec uptests. The manager thread also retains its original privileges in this case.
You should take care that the user you specify has write permissions on your cache file and
status pipe (if you need a status pipe). You should look out for error messages like "permission denied"
and "operation not permitted" to discover permission problems.
.br
.TP
.B strict_setuid=(on|off);
When used together with the run_as option, this option lets you specify that all threads of the
program will run with the privileges of the run_as user. This provides higher security than
the normal run_as
option, but is not always possible. See the run_as option for further discussion.
.br
This option is on by default.
.br
Note that this option has no effect on Non-Linux systems.
.TP
.B paranoid=(on|off);
Normally, pdnsd queries all servers in recursive mode (i.e. instructs servers to query other servers themselves
if possible,
and to give back answers for domains that may not be in its authority), and accepts additional records with information
for servers that are not in the authority of the queried server. This opens the possibility of so-called cache poisoning:
a malicious attacker might set up a dns server that, when queried, returns forged additional records. This way, he might
replace trusted servers with his own ones by making your dns server return bad IP addresses. This option protects
you from cache poisoning by rejecting additional records
that do not describe domains in the queried servers authority space and not doing recursive queries any more.
An exception
to this rule are the servers you specify in your config file, which are trusted.
.br
The penalty is a possible performance decrease, in particular, more queries might be necessary for the same
operation.
.br
You should also notice that there may be other similar security problems, which are essentially problems of
the DNS, i.e.
any "traditional" server has them (the DNS security extensions solve these problems, but are not widely
supported).
One of this vulnerabilities is that an attacker may bombard you with forged answers in hopes that one may match a
query
you have done. If you have done such a query, one in 65536 forged packets will be succesful (i.e. an average packet
count of 32768 is needed for that attack). pdnsd can use TCP for queries,
which has a slightly higher overhead, but is much less vulnerable to such attacks on sane operating systems. Also, pdnsd
chooses random query ids, so that an attacker cannot take a shortcut. If the attacker is able to listen to your network
traffic, this attack is relatively easy, though.
.br
This vulnerability is not pdnsd's fault, and is possible using any conventional
name server (pdnsd is perhaps a little more secured against this type of attacks if you make it use TCP).
.br
The paranoid option is off by default.
.br
.TP
.B ignore_cd=(on|off);
\fINew in version 1.2.8:\fP This option lets you specify that the CD bit of a DNS query will be ignored.
Otherwise pdnsd will reply FORMERR to clients that set this bit in a query.
It is safe to enable this option, as the CD bit refers to 'Checking Disabled'
which means that the client will accept non-authenticated data.
.br
This option is on by default. Turn it off if you want the old behaviour (before version 1.2.8).
.TP
.B scheme_file=\fIstring\fP;
In addition to normal uptests, you may specify that some servers shall only be queried when a certain
pcmcia-cs scheme is active (only under linux). For that, pdnsd needs to know where the file resides that
holds the pcmcia scheme information. Normally, this is either /var/lib/pcmcia/scheme or
/var/state/pcmcia/scheme.
.TP
.B status_ctl=(on|off);
This has the same effect as the \-s command line option: the status control is enabled when
on is specified.
.br
\fIAdded by Paul Rombouts\fP: Note that pdnsd\-ctl allows run-time configuration of pdnsd,
even the IP addesses of the name servers can be changed. If you're not using pdnsd\-ctl and
you want maximum security, you should not enable this option. It is disabled by default.
.TP
.B daemon=(on|off);
This has the same effect as the \-d command line option: the daemon mode is enabled when
on is specified.
.br
Default is off.
.TP
.B tcp_server=(on|off);
tcp_server=on has the same effect as the \-t or \-\-tcp
command-line option: it enables TCP serving.
Similarly, tcp_server=off is like the \-\-notcp command-line option.
.br
Default is on.
.TP
.B pid_file=\fIstring\fP;
This has the same effect as the \-p command line option: you can specify a file that pdnsd
will write its pid into when it starts in daemon mode.
.TP
.B verbosity=\fInumber\fP;
This has the same effect as the \-v command line option: you can set the verbosity of pdnsd's
messages with it. The argument is a number between 0 (few messages) to 3 (most messages).
.TP
.B query_method=(tcp_only|udp_only|tcp_udp|udp_tcp);
This has the same effect as the \-m command line option.
Read the documentation for the command line option on this.
tcp_only corresponds to the to, udp_only to the uo,
tcp_udp to the tu and udp_tcp to the ut
argument of the command line option.
.br
If you use query_method=tcp_udp, it is recommended that you also set the global timeout option to at least twice the longest server timeout.
.TP
.B run_ipv4=(on|off);
This has the same effect as the \-4 or \-6 command line option:
if on is specified, IPv4 support is enabled, and IPv6 support is disabled (if available).
If off is specified, IPv4 will be disabled and IPv6 will be enabled.
For this option to be meaningful, pdnsd needs to be compiled with support for the protocol you choose.
If pdnsd was compiled with both IPv4 and IPv6 support, and you want to include IPv6 addresses
in the configuration file, you will probably need to specify run_ipv4=off first to
ensure that the IPv6 addresses are parsed correctly.
.TP
.B debug=(on|off);
This has the same effect as the \-g command line option: the debugging messages are enabled when
on is specified.
.TP
.B ctl_perms=\fInumber\fP;
This option allows you to set the file permissions that the pdnsd status control socket will have. These
are the same as file permissions. The owner of the file will be the run_as user, or, if none is specified,
the user who started pdnsd. If you want to specify the permissions in octal (as usual), don't forget
the leading zero (0600 instead of 600!). To use the status control, write access is needed. The default
is 0600 (only the owner may read or write).
.br
Please note that the socket is kept in the cache directory, and that the cache directory permissions
might also need to be adjusted. Please ensure that the cache directory is not writeable for untrusted
users.
.TP
.B proc_limit=\fInumber\fP;
With this option, you can set a limit on the pdnsd threads that will be active simultaneously. If
this number is exceeded, queries are queued and may be delayed some time.
See also the procq_limit option.
.br
The default for this option is 40.
.TP
.B procq_limit=\fInumber\fP;
When the query thread limit proc_limit is exceeded, connection attempts to pdnsd will be queued.
With this option, you can set the maximum queue length.
If this length is also exceeded, the incoming queries will be dropped.
That means that tcp connections will be closed and udp queries will just be dropped, which
will probably cause the querying resolver to wait for an answer until it times out.
.br
See also the proc_limit option. A maximum of proc_limit+procq_limit
query threads will exist at any one time (plus 3 to 6 threads that will always
be present depending on your configuration).
.br
The default for this option is 60.
.TP
.B tcp_qtimeout=\fItimespec\fP;
This option sets a timeout for tcp queries. If no full query has been received on a tcp connection
after that time has passed, the connection will be closed. The default is set using the
\-\-with\-tcp\-qtimeout option to configure.
.TP
.B par_queries=\fInumber\fP;
This option used to set the maximum number of remote servers that would be queried simultaneously,
for every query that pdnsd receives.
.br
Since version 1.1.11, the meaning of this option has changed slightly.
It is now the increment with which the number of parallel queries is
increased when the previous set of servers has timed out.
For example, if we have a list \fIserver1, server2, server3,\fP etc. of available servers
and par_queries=2, then pdnsd will first send queries to \fIserver1\fP and \fIserver2\fP,
and listen for responses from these servers.
.br
If these servers do not send a reply within their timeout period, pdnsd will send additional
queries to \fIserver3\fP and \fIserver4\fP, and listen for responses from
\fIserver1, server2, server3\fP and \fIserver4\fP, and so on until a useful reply is
received or the list is exhausted.
.br
In the worst case there will be pending queries to all the servers in the list of available servers.
We may be using more system resources this way (but only if the first servers in the list
are slow or unresponsive), but the advantage is that we have a greater chance of catching a reply.
After all, if we wait longer anyway, why not for more servers.
.br
See also the explanation of the global timeout option below.
.br
1 or 2 are good values for this option.
The default is set at compile time using the \-\-with\-par\-queries option to configure.
.TP
.B timeout=\fItimespec\fP;
This is the global timeout parameter for dns queries.
This specifies the minimum period of time pdnsd will wait after sending the
first query to a remote server before giving up without having
received a reply. The timeout options in the configuration file are
now only minimum timeout intervals. Setting the global timeout option
makes it possible to specify quite short timeout intervals in the
server sections (see below). This will have the effect that pdnsd will start
querying additional servers fairly quickly if the first servers are
slow to respond (but will still continue to listen for responses from
the first ones). This may allow pdnsd to get an answer more quickly in
certain situations.
.br
If you use query_method=tcp_udp it is recommended that
you make the global timeout at least twice as large as the largest
server timeout, otherwise pdnsd may not have time to try a UDP query
if a TCP connection times out.
.br
Default value is 0.
.TP
.B randomize_recs=(on|off);
If this option is turned on, pdnsd will randomly reorder the cached records of one type
when creating an answer. This supports round-robin DNS schemes and increases fail
safety for hosts with multiple IP addresses, so this is usually a good idea.
.br
On by default.
.TP
.B query_port_start=(\fInumber\fP|none);
If a number is given, this defines the start of the port range used for queries of pdnsd. The
value given must be >= 1024. The purpose of this option is to aid certain firewall
configurations that are based on the source port. Please keep in mind that another application
may bind a port in that range, so a stateful firewall using target port and/or process uid may
be more effective. In case a query start port is given pdnsd uses this port as the first port of a
specified port range (see query_port_end) used for queries.
pdnsd will try to randomly select a free port from this range as local port for the query.
.br
To ensure that there are enough ports for pdnsd to use, the range between query_port_start and
query_port_end should be adjusted to at least (par_queries * proc_limit).
A larger range is highly recommended for security reasons, and also because other applications may
allocate ports in that range. If possible, this range should be kept out of the space
that other applications usually use.
.br
The default for this option is 1024. Together with the default value of query_port_end,
this makes it the hardest for an attacker to guess the source port used by the pdnsd resolver.
If you specify none here, pdnsd will let the kernel choose the source port, but
this may leave pdnsd more vulnerable to an attack.
.TP
.B query_port_end=\fInumber\fP;
Used if query_port_start is not none. Defines the last port of the range started by query_port_start
used for querys by pdnsd. The default is 65535, which is also the maximum legal value for this option.
For details see the description of query_port_start.
.TP
.B delegation_only=\fIstring\fP;
\fIAdded by Paul Rombouts\fP: This option specifies a "delegation-only" zone.
This means that if pdnsd receives a query for a name that is in a
subdomain of a "delegation-only" zone but the remote name server
returns an answer with an authority section lacking any NS RRs for
subdomains of that zone, pdnsd will answer NXDOMAIN (unknown domain).
This feature can be used for undoing the undesired effects of DNS
"wildcards". Several "delegation-only" zones may be specified together.
If you specify root servers in a server section it is
important that you set root_server=on in such a section.
.br
Example:

delegation_only="com","net";

This feature is off by default. It is recommended that you only use
this feature if you actually need it, because there is a risk that
some legitimate names will be blocked, especially if the remote
name servers queried by pdnsd return answers with empty authority
sections.
.TP
.B ipv4_6_prefix=\fIstring\fP;
This option has the same effect as the \-i command-line option.
When pdnsd runs in IPv6 mode, this option specifies the prefix pdnsd uses to convert IPv4 addresses in
the configuration file (or addresses specified with pdnsd\-ctl)
to IPv6-mapped addresses.
The string must be a valid IPv6 address. Only the first 96 bits are used.
Note that this only effects the parsing of IPv4 addresses listed after this option.
.br
The default is "::ffff.0.0.0.0".
.TP
.B use_nss=(on|off);
If this option is turned on, pdnsd will call initgroups() to set up the group access list,
whenever pdnsd changes its user and group id (see run_as option).
There is a possible snag, though, if initgroups() uses NSS (Name Service Switch) and
NSS in turn uses DNS. In such a case you may experience lengthy timeouts and stalls.
By setting use_nss=off, you can disable the initgroups() call
(only possible in versions 1.2.5 and later).
.br
This option was contributed by Jan-Marek Glogowski.
.br
On by default.
.TP
.B udpbufsize=\fInumber\fP;
\fINew in version 1.2.9:\fP
This option sets the upper limit on the size of UDP DNS messages. The default is 1024.
.br
See also the edns_query server option below.

.SS server Section
Each server section specifies a set of name servers that pdnsd should try to get
resource records or authoritative name server information from. The servers are
queried in the order of their appearance (or parallel to a limited extend).
If one fails, the next one is taken and so on.
.br
You probably want to specify  the dns server in your LAN, the caching dns servers
of your internet provider or even a list of root servers in one or more server sections.
.br
The supported options in this section are:

.TP
.B label=\fIstring\fP;
Specify a label for the server section. This can be used to refer to this section
when using pdnsd\-ctl, the pdnsd control utility.
.br
You can give several server sections the same label, but if you want to change the addresses
of a server section (see \fBip\fP option below) during run-time with
"pdnsd\-ctl\ server\ \fIlabel\fP\ up\ \fIdns1\fP,\fIdns2\fP,...",
the label must be unique.
.TP
.B ip=\fIstring\fP;
Give the IP (the address, \fInot\fP the host name) of the server.
.br
Multiple IP addresses can be given per server section.
This can be done by entering multiple lines of the form ip=\fIstring\fP;
or a single line like this:

ip=\fIstring\fP,\fIstring\fP,\fIstring\fP;

IP addresses do not have to be specified in the configuration file.
A server section without IP addresses will remain inactive until it is assigned
one or more addresses with pdnsd\-ctl,
the pdnsd control utility.
.br
If pdnsd has been compiled with both IPv4 and IPv6 support, any IPv6 addresses you specify
here will be skipped with a warning message, unless pdnsd is running in IPv6 mode.
Thus, unless pdnsd was compiled to startup in IPv6 mode by default, you need to use the
command-line option \-6 or set run_ipv4=off
first (see global section) in order to ensure
that IPv6 addresses are parsed correctly.
.br
If pdnsd is running in IPv6 mode and you specify an IPv4 address here,
it will automatically be mapped to an IPv6 address.
.TP
.B file=\fIstring\fP;
\fINew in version 1.2:\fP This option allows you to give the name of a resolv.conf-style file.
Of the lines beginning with the nameserver keyword, the second field will be parsed as an
IP address, as if it were specified with the ip= option. The remaining lines will be ignored.
If the contents of the file changes while pdnsd is running, you can make pdnsd aware of the changes through the
use of pdnsd\-ctl, the pdnsd control utility.
This is usually most conveniently done by placing the command "pdnsd\-ctl\ config" in a script
that is automatically run whenever the DNS configuration changes.
.br
For example, suppose you have a ppp client that writes the DNS configuration for your ISP to the file
/etc/ppp/resolv.conf and runs the script /etc/ppp/ip-up when a new
connection is established. One way of ensuring that pdnsd is automatically reconfigured is to
add a server section in the config file with file=/etc/ppp/resolv.conf and to
add the command "pdnsd\-ctl\ config" to /etc/ppp/ip-up.
.TP
.B port=\fInumber\fP;
Give the port the remote name server listens on. Default is 53 (the official
dns port)
.TP
.B uptest=(ping|none|if|dev|diald|exec|query);
Determine the method to check whether the server is available. Currently
defined methods are:
.IP
\(bu \fBping\fP: Send an ICMP_ECHO request to the server. If it doesn't respond
within the timeout, it is regarded to be unavailable until the next probe.
.IP
\(bu \fBnone\fP: The availability status is not changed, only the time stamp is updated.
.IP
\(bu \fBif\fP: Check whether the interface (specified in the interface= option) is
existent, up and running. This currently works for all "ordinary"
network interfaces, interfaces that disappear when down (e.g. ppp?),
and additionally for Linux isdn interfaces (as of kernel 2.2). Note that
you need a /dev/isdninfo device file (major#45, minor#255), or the
isdn uptest will always fail.
.IP
\(bu \fBdev\fP and \fBdiald\fP: Perform an if uptest, and, if that
was succesful, additionally check whether a program is running that
has locked a given (modem-) device. The needed parameters are an interface (specified as for the if
uptest, e.g. "ppp0") and a device relative to /dev (e.g.
"modem" for /dev/modem specified using the device= option.
pdnsd will then look for a pid file for the given interface in /var/lock (e.g.
/var/run/ppp0.pid) and for a lockfile for the given device (e.g. /var/lock/LCK..modem),
and then test whether the locking process is the process that created the pid file and this process is still
alive. If this is the case, the normal if uptest is executed for the given interface.
.br
The dev option is for pppd dial-on-demand, diald is the same for diald users.
.IP
\(bu \fBexec\fP: Executes a given command in the /bin/sh shell
(as /bin/sh \-c <command>)
and evaluates the result (the return code of the last command) in the shell's way of handling return codes,
i.e. 0 indicates success, all other indicate failure. The shell's process name will be
uptest_sh. The command is given with the uptest_cmd option (see below).
For secuity issues, also see that entry.
.IP
\(bu \fBquery\fP: \fINew in version 1.2:\fP
This works like the ping test, except it sends an (empty) DNS query to the remote server.
If the server sends a well-formed response back within the timeout period (except SERVFAIL),
it will be regarded as available.
This test is useful if a remote server does not respond to ICMP_ECHO requests at all,
which unfortunately is quite common these days.
It can also happen that a remote server is online but ignores empty DNS queries.
Then you will need the set the query_test_name option (see below).
In many cases this test will be a more reliable indicator of availability
than the ones mentioned before.

The default value is \fBnone\fP.

\fBNOTE\fP: If you use on-demand dialing, use none, if,
dev, diald or exec,
since ping or query will send packets
in the specified interval and the interface will thus frequently dial!
.TP
.B ping_timeout=\fInumber\fP;
Sets the timeout for the ping test in tenths of seconds
(this unit is used for legacy reasons; actually the current implementation is
only accurate to a second).
.br
The default is 600 (one minute).
.TP
.B ping_ip=\fIstring\fP;
The IP address for the ping test. The default is the IP of the name server.
.TP
.B query_test_name=\fIstring\fP;
\fINew in version 1.2.9:\fP
Sets the name to be queried when using uptest=query availability test.

changelog.md
##2020.05.07##
* 增加zerotier 1.4.6内网穿透
* openssl更新1.1.1g
* 取消32M FLASH的重启选项,默认使用MTD UNLOCK方式重启
* trojan升级到1.15.1,集成到源码动态编译
* MT7615增加5.0.4.0驱动
* SSP增加关键字过滤,修复V2 HTTP
* SSP界面增加节点编辑,复制功能
* smartdns升级到最新源码
* 同步hanwckf的最新源码
* 自动编译增加红米2100和小米2100固件

##往期更新日志##
*https://github.com/chongshengB/rt-n56u/blob/master/changelog-old.md


trunk/user/pdnsd/pdnsd.sh
#!/bin/sh
dns_enable=`nvram get dns_enable`
dns_server=`nvram get dns_server`
dns_server_bind=`nvram get dns_server_bind`
dns_server_port=`nvram get dns_server_port`
PDNSD_BIN="/usr/bin/pdnsd"
ss_dns=$2
pdnsd_genconfig() {	
  if [ ! -f /tmp/dnscache/pdnsd.cache ]; then
    mkdir -p /tmp/dnscache
    echo -ne "pd13\000\000\000\000" > /tmp/dnscache/pdnsd.cache
    chown -R nobody.nogroup /tmp/dnscache
	fi
if [ ! -z $ss_dns ];then
dns_server="$dns_server,$ss_dns"
fi
	cat > /tmp/dnscache.conf <<EOF
global {
    perm_cache=512;        # dns缓存大小，单位KB，建议不要写的太大
    cache_dir="/tmp/dnscache";     # 缓存文件的位置
    pid_file = /tmp/dnscache.pid;
    server_ip = $dns_server_bind;        # pdnsd监听的网卡，0.0.0.0是全部网卡
    server_port=$dns_server_port;           # pdnsd监听的端口，不要和别的服务冲突即可
    status_ctl = on;
    paranoid=on;                  # 二次请求模式，如果请求主DNS服务器返回的是垃圾地址，就向备用服务器请求
    query_method=tcp_udp;      
    neg_domain_pol = off;  
    par_queries = 400;          # 最多同时请求数
    min_ttl = 1h;               # DNS结果最短缓存时间
    max_ttl = 1w;               # DNS结果最长缓存时间
    timeout = 10;               # DNS请求超时时间，单位秒
}

server {  
    label = "routine";         
    ip = $dns_server;     # 这里为主要上级 dns 的 ip 地址，建议填写一个当地最快的DNS地址  
    timeout = 5;              # DNS请求超时时间
    reject = 74.125.127.102,  # 以下是脏IP，也就是DNS污染一般会返回的结果，如果收到如下DNS结果会触发二次请求（TCP协议一般不会碰到脏IP）
        74.125.155.102,  
        74.125.39.102,  
        74.125.39.113,  
        209.85.229.138,  
        128.121.126.139,  
        159.106.121.75,  
        169.132.13.103,  
        192.67.198.6,  
        202.106.1.2,  
        202.181.7.85,  
        203.161.230.171,  
        203.98.7.65,  
        207.12.88.98,  
        208.56.31.43,  
        209.145.54.50,  
        209.220.30.174,  
        209.36.73.33,  
        211.94.66.147,  
        213.169.251.35,  
        216.221.188.182,  
        216.234.179.13,  
        243.185.187.39,  
        37.61.54.158,  
        4.36.66.178,  
        46.82.174.68,  
        59.24.3.173,  
        64.33.88.161,  
        64.33.99.47,  
        64.66.163.251,  
        65.104.202.252,  
        65.160.219.113,  
        66.45.252.237,  
        69.55.52.253,  
        72.14.205.104,  
        72.14.205.99,  
        78.16.49.15,  
        8.7.198.45,  
        93.46.8.89,  
        37.61.54.158,  
        243.185.187.39,  
        190.93.247.4,  
        190.93.246.4,  
        190.93.245.4,  
        190.93.244.4,  
        65.49.2.178,  
        189.163.17.5,  
        23.89.5.60,  
        49.2.123.56,  
        54.76.135.1,  
        77.4.7.92,  
        118.5.49.6,  
        159.24.3.173,  
        188.5.4.96,  
        197.4.4.12,  
        220.250.64.24,  
        243.185.187.30,  
        249.129.46.48,  
        253.157.14.165;  
    reject_policy = fail;  
}

server {  
    label = "special";                  # 这个随便写  
    ip = 208.67.222.222,208.67.220.220; # 这里为备用DNS服务器的 ip 地址  
    port = 53;                        # 推荐使用53以外的端口（DNS服务器必须支持） 
    proxy_only = on;
    timeout = 5;  
}  

source {
	owner=localhost;
//	serve_aliases=on;
	file="/etc/hosts";
}

rr {
	name=localhost;
	reverse=on;
	a=127.0.0.1;
	owner=localhost;
	soa=localhost,root.localhost,42,86400,900,86400,86400;
}
EOF

chmod -R 744 /tmp/dnscache.conf
	echo "Start DNS Cache"
	logger -t "pdnsd" "启动DNS加速。"
}

dns_close() {
  cat /tmp/dnscache.pid | xargs kill -9
  revert_dns
  echo "Stop DNS Cache"
  logger -t "pdnsd" "关闭DNS加速。"
}

change_dns() {
sed -i '/no-resolv/d' /etc/storage/dnsmasq/dnsmasq.conf
sed -i '/server=127.0.0.1/d' /etc/storage/dnsmasq/dnsmasq.conf
cat >> /etc/storage/dnsmasq/dnsmasq.conf << EOF
no-resolv
server=127.0.0.1#$dns_server_port
EOF

/sbin/restart_dhcpd
}

revert_dns() {
sed -i '/no-resolv/d' /etc/storage/dnsmasq/dnsmasq.conf
sed -i '/server=127.0.0.1/d' /etc/storage/dnsmasq/dnsmasq.conf
/sbin/restart_dhcpd
}

dns_start(){
 # if [ $dns_enable -eq 1 ];  then
  pdnsd_genconfig
  if [ ! -f "/var/pdnsddns" ]; then
  ln -sf /usr/bin/pdnsd /var/pdnsddns
  fi
/var/pdnsddns -c /tmp/dnscache.conf -d
    change_dns
    
 #fi
}

case $1 in
start)
	dns_start
	;;
stop)
	dns_close
	;;
*)
	echo "check"
	;;
esac
